### PR TITLE
solver: fix dependency resolution for self-referential extras with duplicate dependencies

### PR DESCRIPTION
**Pull request by radoering**
# Pull Request Check List

Resolves: #10434

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

## Summary by Sourcery

Fix solver dependency resolution for self-referential extras containing duplicate dependencies and extend test fixtures to cover merged extra markers

Enhancements:
- Use found_extras instead of dependency.extras when resolving overlapping markers for non-root packages to handle duplicate extra dependencies correctly

Tests:
- Add merge_extras parameter to the test package factory to support merging markers of duplicate optional dependencies
- Parametrize test_solver_resolves_self_referential_extras on merge_extras and include duplicate dependent extras in the test data

### CODE REVIEW

File: tests/puzzle/test_solver.py [hunk 1]
```diff
@@ -580,20 +581,23 @@ def test_solver_resolves_self_referential_extras(
...
             str(package.version),
             extras={
                 "download": ["download-package"],
+                "download2": ["download-package"],  # same package as download
                 "install": ["install-package"],
                 "py38": ["py38-package ; python_version == '3.8'"],
                 "py310": ["py310-package ; python_version > '3.8'"],
-                "all": ["a[download,install]"],
+                "all": ["a[download,download2,install]"],
                 "py": ["a[py38,py310]"],
```
[the remark this answers is not in the document]
Comment (radoering) [reply tests/puzzle/test_solver.py:597]: out of scope

File: tests/puzzle/test_solver.py [hunk 2]
```diff
@@ -580,20 +581,23 @@ def test_solver_resolves_self_referential_extras(
...
                 "nested": ["a[all]"],
```
[the remark this answers is not in the document]
Comment (radoering) [reply tests/puzzle/test_solver.py:598]: out of scope

[the remark this answers is not in the document]
Comment (radoering) [reply tests/puzzle/test_solver.py:598]: nevertheless a good proposal, follow up: #10489
